### PR TITLE
Support NRT and MSV in `required` and `nullable`

### LINKED ISSUE #1111: Add support for handling of non-nullable reference types in `required`.
For non-nullable reference types, a resource attribute should be included in the OpenAPI `required` property if model state validation is enabled, and should be excluded otherwise.

The following table displays a comprehensive overview of the behavior of `required`, `nullable` in an OpenAPI schema, depending on the property type, usage of `[Required]`, configuration of NRT, and configuration of model state validation.

177394179-2943249b-f5f1-4a72-adab-1c1d7ae85fc6

### LINKED ISSUE #1242: Update JsonApiDotNetCoreExampleClient
#### DESCRIPTION
The included swagger.json file is outdated. It should be refreshed, and the example should be expanded to cover all JSON:API endpoints.

### PR DESCRIPTION
**Pull request by maurei**
This PR introduces improved support for `required` and `nullable` properties in OpenAPI schemas, by taking into account enabled/disabled nullable reference types and enabled/disabled model state validation. The table in [Google Sheets](https://docs.google.com/spreadsheets/d/1er4YxE1An9GEyKeHAGwTN0FlDrbOWi0PNdtw-eIM8wA/edit?usp=sharing) (copy below) supersedes the original table from #1111.

Includes:
- Use the native `NullabilityInfoContext` to check nullability information rather than relying on Swashbuckles work-around (supersedes #1167).
- Take into account if model state validation is enabled/disabled when generating OpenAPI schemas.
- Small tweak in `OpenApiTestContext` to allow for retrieving a swagger file without having to write it to disk.

Closes #1111
Closes #1242

image

#### QUALITY CHECKLIST
- [x] Changes implemented in code
- [x] Complies with our [contributing guidelines](https://github.com/json-api-dotnet/JsonApiDotNetCore/blob/master/.github/CONTRIBUTING.md)
- [x] Adapted tests
- [x] Documentation updated

### CODE REVIEW

**Review by bkoelman — CHANGES_REQUESTED** [review]
Very nice job! The implementation is straightforward and it's easy to verify the tests against what's in https://user-images.githubusercontent.com/9339132/177394179-2943249b-f5f1-4a72-adab-1c1d7ae85fc6.png. The code looks clean with good naming and structure.

The used models remind me of the [Cow and Chicken series](https://www.youtube.com/watch?v=yj3bpF5VPi4) that my younger brother always watched! :)

**Comment by bkoelman** [comment]
I took the liberty to merge in #1186, I hope you don't mind. However, I was surprised that _running the tests_ changed two swagger.g.json files on disk, _without_ any of the tests failing. That doesn't seem right: when an OAS file changes unexpectedly, something definitely should fail.

File: test/OpenApiTests/SchemaProperties/NullableReferenceTypesDisabled/ModelStateValidationEnabledTests.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    public ModelStateValidationEnabledTests(
+        OpenApiTestContext<SchemaPropertiesStartup<NullableReferenceTypesDisabledDbContext>, NullableReferenceTypesDisabledDbContext> testContext)
+    {
+        _testContext = testContext;
+
+        testContext.UseController<ChickensController>();
+    }
+
+    [Fact]
+    public async Task Resource_when_ModelStateValidation_is_enabled_produces_expected_required_property_in_schema()
```
[the remark this answers is not in the document]
Comment (maurei) [reply test/OpenApiTests/SchemaProperties/NullableReferenceTypesDisabled/ModelStateValidationEnabledTests.cs:22]: For the schemas in this test it is relevant that their generation is driven by resources specifically, it was my intent to convey that. With the "when" clause removed that becomes clearer, and I agree that clause can be dropped.

I believe `property` is fine. The term is commonly used for key-value pairs in JSON objects, and it is frequently used in the OpenAPI spec too. It is also consistent with the test folder name.

**Comment by maurei** [comment]
Review item's I've reacted to are processed in my latest commit. The other ones are related to something I'm still investigating and will follow

File: test/OpenApiTests/SchemaProperties/NullableReferenceTypesDisabled/ModelStateValidationEnabledTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bkoelman) [reply test/OpenApiTests/SchemaProperties/NullableReferenceTypesDisabled/ModelStateValidationEnabledTests.cs]: You're right that the terms "property" and "element" are both widely used in JSON and OpenAPI content online. Apparently, this is a case of using multiple terms for the same thing. I prefer we'd use "element" for two reasons:
1. It takes away any ambiguity with .NET properties.
2. The code inside tests refer to "element" for them, because they are typed as `JsonElement`. So using that same terminology in the test name conveys it's about the same thing.

[the remark this answers is not in the document]
Comment (bkoelman) [reply test/OpenApiTests/SchemaProperties/NullableReferenceTypesDisabled/ModelStateValidationEnabledTests.cs]: Related to this, the test named "Produces_expected_nullable_properties_in_schema_for_resource" adds to the terminology confusion. It actually asserts that _elements_ are added, which are named "properties".

[the remark this answers is not in the document]
Comment (maurei) [reply test/OpenApiTests/SchemaProperties/NullableReferenceTypesDisabled/ModelStateValidationEnabledTests.cs]: Here `element` and `property` are not interchangeable. The `property` choice is in line with [Microsoft.OpenApi.Models.OpenApiSchema#Properties](https://github.com/microsoft/OpenAPI.NET/blob/760445cf88cdb6aff00a6ca3a0fd0827f45f91db/src/Microsoft.OpenApi/Models/OpenApiSchema.cs#L165), which is based on the terminology from the corresponding [JSON Schema spec item](https://datatracker.ietf.org/doc/html/draft-fge-json-schema-validation-00#section-5.4.4.1), and `element` is not because in that doc it occurs only in the context of arrays. 

The little (imo non-existent) ambiguity with .NET properties here isn't a strong enough reason to break with the laid out  terminology. Also consider the naming of tangent properties like `AdditionalPropertiesAllowed`, `MaxProperties`, `MinProperties`. See also [property set](https://datatracker.ietf.org/doc/html/draft-fge-json-schema-validation-00#section-2) where `element` and `property` are used in a non-interchangeable manner.


> So using that same terminology in the test name conveys it's about the same thing.

To me they clearly do not: the test name contains a high-level OAS related term (required property), whereas `element` refers to a low-level technical detail (`JsonElement` clr type), hence _not_ the same. I'll take this confusion out by making no reference to `JsonElement`

File: test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesDisabled/NullabilityTests.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+        propertyInfo = propertyInfos.FirstOrDefault(property => property.Name == nameof(ChickenAttributesInResponse.NameOfCurrentFarm));
+        propertyInfo.Should().BeNullable();
+
+        propertyInfo = propertyInfos.FirstOrDefault(property => property.Name == nameof(ChickenAttributesInResponse.Age));
+        propertyInfo.Should().BeNonNullable();
+
+        propertyInfo = propertyInfos.FirstOrDefault(property => property.Name == nameof(ChickenAttributesInResponse.TimeAtCurrentFarmInDays));
+        propertyInfo.Should().BeNullable();
+    }
```
[the remark this answers is not in the document]
Comment (maurei) [reply test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesDisabled/NullabilityTests.cs:26]: Here `element` and `property` are not interchangeable. The choice for `property` comes from [Microsoft.OpenApi.Models.OpenApiSchema#Properties](https://github.com/microsoft/OpenAPI.NET/blob/760445cf88cdb6aff00a6ca3a0fd0827f45f91db/src/Microsoft.OpenApi/Models/OpenApiSchema.cs#L165), which is in accordance with terminology defined in the [JSON Schema spec](https://datatracker.ietf.org/doc/html/draft-fge-json-schema-validation-00#section-2) despite of little (imo non-existent) risk of ambiguity, whereas `element` isn't, because in this doc "element" only occurs in the context of collections. Also consider the naming of tangent properties like `AdditionalPropertiesAllowed`, `MaxProperties`, `MinProperties`.

1) 

So unless there is a really strong reason, I don't want to deviate from the standard term here.


Technically, the value in the key-value pair `"required": [ "prop1", "prop2", "prop3" ]` is defined as a [property set](https://datatracker.ietf.org/doc/html/draft-fge-json-schema-validation-00#section-2) where the set consists of _elements_. I.e. to refer to this key-value pair as a whole as _element_ mixes up that terminology. 

so `Produces_expected_required_property_in_schema_for_resource` could be improved slightly to `Produces_expected_required_property_set_in_schema_for_resource`



> Related to this, the test named "Produces_expected_nullable_properties_in_schema_for_resource" adds to the terminology confusion. It actually asserts that elements are added, which are named "properties".

"Nullable properties" denotes is high-level functional behaviour (which I think is good to have in the test name), whereas "element" in the test body refers to a technical implementation detail (CLR type `JsonElement`) so I don't think these are confusing. I can make it into "JsonElement", or leave the type name out of the variable.

File: test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesEnabled/NullabilityTests.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+        propertyInfo = propertyInfos.FirstOrDefault(property => property.Name == nameof(CowAttributesInResponse.NameOfPreviousFarm));
+        propertyInfo.Should().BeNullable();
+
+        propertyInfo = propertyInfos.FirstOrDefault(property => property.Name == nameof(CowAttributesInResponse.Age));
+        propertyInfo.Should().BeNonNullable();
+
+        propertyInfo = propertyInfos.FirstOrDefault(property => property.Name == nameof(CowAttributesInResponse.TimeAtCurrentFarmInDays));
+        propertyInfo.Should().BeNullable();
+    }
```
[the remark this answers is not in the document]
Comment (maurei) [reply test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesEnabled/NullabilityTests.cs:26]: 1,2,3,5: OK

About 4: We shouldn’t inspect the `JsonPropertyAttribute` here because it addresses a completely different testing concern that doesn’t belong to this test suite.

The concern in this test suite is the CLR nullability in the generated source code, i.e. it’s the equivalent of the `GeneratedTypeTests` from the existing test suite. This nullability is in no way affected by `JsonPropertyAttribute` (Such tests are questionable because they don’t test our own code, but we decided to occasionally do so anyway because we’re in very new waters).

Instead, testing `JsonPropertyAttribute` addresses the concern of how `JsonApiClient` behaves at runtime during serialization. But by inspecting this attribute directly it does so only poorly because there are many other factors at play: `DefaultValueHandling`, `NullValueHandling`, how the `JsonApiClient` is called, etc. Instead this should be tested jointly in a behavioural test, which probes these attributes in their relevant context. That’s what happens in `RequiredAttributesTests`.

[the remark this answers is not in the document]
Comment (maurei) [reply test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesEnabled/NullabilityTests.cs:26]: I’m currently working out a more rigorous test suite that will replace RequiredAttributesTests with a similar suite that covers more scenarios. This will further capture the runtime behaviour of `JsonApiClient`. 

For these changes I will make a PR into this branch to allow for better reviewability

File: test/OpenApiTests/SchemaProperties/NullableReferenceTypesDisabled/ModelStateValidationEnabledTests.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+using System.Text.Json;
+using FluentAssertions;
+using TestBuildingBlocks;
+using Xunit;
+
+namespace OpenApiTests.SchemaProperties.NullableReferenceTypesDisabled;
+
+public sealed class ModelStateValidationEnabledTests
```
[the remark this answers is not in the document]
Comment (maurei) [reply test/OpenApiTests/SchemaProperties/NullableReferenceTypesDisabled/ModelStateValidationEnabledTests.cs:8]: Initially I thought it could be neglected for that precise reason. 

But now I think it is useful and complete to demonstrate that those attributes which cannot be omitted during a POST request, can be omitted during a PATCH request.

File: test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesDisabled/RequiredAttributesTests.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                {
+                    Age = 1
+                }
+            }
+        };
+
+        Func<Task> action = async () => await ApiResponse.TranslateAsync(async () => await apiClient.PatchChickenAsync(1, requestDocument));
+
+        // Assert
+        await action.Should().ThrowAsync<JsonSerializationException>();
```
[the remark this answers is not in the document]
Comment (maurei) [reply test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesDisabled/RequiredAttributesTests.cs:111]: This will be superseded by a test in #1231.

**Comment by maurei** [comment]
> I took the liberty to merge in #1186, I hope you don't mind. However, I was surprised that _running the tests_ changed two swagger.g.json files on disk, _without_ any of the tests failing. That doesn't seem right: when an OAS file changes unexpectedly, something definitely should fail.

Haven't been able to reproduce this.  Compared two fresh clones of this branch where in only one of them I ran the tests

File: src/JsonApiDotNetCore.OpenApi/SwaggerComponents/ResourceFieldObjectSchemaBuilder.cs [hunk 6]
```diff
@@ -212,18 +224,12 @@ private static bool IsRelationshipInResponseType(Type relationshipSchemaType)
     {
         Type relationshipSchemaOpenType = relationshipSchemaType.GetGenericTypeDefinition();
 
-        return RelationshipInResponseOpenTypes.Contains(relationshipSchemaOpenType);
+        return RelationshipSchemaInResponseOpenTypes.Contains(relationshipSchemaOpenType);
     }
 
-    private static bool IsDataPropertyNullable(Type type)
+    private static bool IsDataPropertyNullableInRelationshipSchemaType(Type relationshipSchemaType)
```
[the remark this answers is not in the document]
Comment (maurei) [reply src/JsonApiDotNetCore.OpenApi/SwaggerComponents/ResourceFieldObjectSchemaBuilder.cs:227]: It's true that relationships require equally thorough test suites as we have for attributes for both server-side OAS generation (like in this PR) as well as client-side request behaviour (like in #1231).  I didn't really realise that until later because the code changes here `JsonApiDotNetCore.OpenApi` exclusively target support for `nullable`/`required` for attributes, not relationships.

So, indeed, this is the logical next step after the current work on attributes is finished.

[the remark this answers is not in the document]
Comment (maurei) [reply src/JsonApiDotNetCore.OpenApi/SwaggerComponents/ResourceFieldObjectSchemaBuilder.cs:227]: For equally rigorous relationship tests, we'll need `Chicken`/`Cow` models that have 4 to 8 relationship properties.

I would like to propose to _not_ use `Chicken`/`Cow` for this, because this will cause the overviews in #1111 and #1231 to explode and be hard to follow, and same for the associated test suites. Especially the client tests will become very verbose and complicated. Eg all attribute related request body declarations will need a lot of redundant relationship data, and conversely.

Instead, I would like to propose to have `SchemaProperties/AttributesObject` tests and `SchemaProperties/RelationshipObjects` tests with dedicated models. Relationships and attributes behave differently so can be seen as separate concerns. This will keep it grouped by those concern.

Given that it wouldn't overlap, it could go in a separate PR dedicated for relationships, and keep this one for attributes

**Comment by maurei** [comment]
@bkoelman there are tests failing due to timeouts unrelated to this PR. It seems to happens during the test cleanup phase which you recently worked on. Do you have any idea what could be causing this?

See for example https://ci.appveyor.com/project/json-api-dotnet/jsonapidotnetcore/builds/45792318/job/l2grxineggemqslq#L317

File: test/OpenApiTests/SchemaProperties/NullableReferenceTypesDisabled/ModelStateValidationEnabledTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (bkoelman) [reply test/OpenApiTests/SchemaProperties/NullableReferenceTypesDisabled/ModelStateValidationEnabledTests.cs]: I agree we should stick with the OpenAPI terminology in test names. The JSON elements are an implementation detail, which only applies _when choosing the JSON file format_. It makes no sense for the YAML file format. This matches the guideline that tests should state a rule, whose implementation is based on an example (using JSON in this case).

File: test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesEnabled/NullabilityTests.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bkoelman) [reply test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesEnabled/NullabilityTests.cs]: Closing, all concerns have been addressed.

File: test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesDisabled/NullabilityTests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bkoelman) [reply test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesDisabled/NullabilityTests.cs]: The suggestion is outdated, however, the underlying concerns have been addressed.

File: test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesDisabled/RequiredAttributesTests.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bkoelman) [reply test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesDisabled/RequiredAttributesTests.cs]: Closing, this has been addressed.

File: src/JsonApiDotNetCore.OpenApi/SwaggerComponents/ResourceFieldObjectSchemaBuilder.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (bkoelman) [reply src/JsonApiDotNetCore.OpenApi/SwaggerComponents/ResourceFieldObjectSchemaBuilder.cs:217]: @maurei It looks like this is now implemented for create/update resource endpoints (see OpenApiTests...NullabilityTests/RequiredTests and OpenApiClientTests...CreateResourceTests/UpdateResourceTests), but not yet for relationship endpoints. Can you confirm?

File: test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesDisabled/RequiredAttributesTests.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+using JsonApiDotNetCore.Middleware;
+using Microsoft.Net.Http.Headers;
+using Newtonsoft.Json;
+using OpenApiClientTests.SchemaProperties.NullableReferenceTypesDisabled.GeneratedCode;
+using TestBuildingBlocks;
+using Xunit;
+
+namespace OpenApiClientTests.SchemaProperties.NullableReferenceTypesDisabled;
+
+public sealed class RequiredAttributesTests
```
[the remark this answers is not in the document]
Comment (bkoelman) [reply test/OpenApiClientTests/SchemaProperties/NullableReferenceTypesDisabled/RequiredAttributesTests.cs]: Closing, it looks like all cases have been addressed.

File: src/JsonApiDotNetCore.OpenApi/SwaggerComponents/ResourceFieldObjectSchemaBuilder.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (bkoelman) [reply src/JsonApiDotNetCore.OpenApi/SwaggerComponents/ResourceFieldObjectSchemaBuilder.cs:216]: Discussed offline: these use the same OAS schemas, so not relevant to test separately.

**Review by bkoelman — APPROVED** [review]
